pil-analyzer/src/condenser.rs

```diff
@@ -940,16 +940,32 @@ fn try_value_to_expression<T: FieldElement>(value: &Value<'_, T>) -> Result<Expr
         }
         .into(),
         Value::Closure(c) => try_closure_to_expression(c)?,
-        Value::TypeConstructor(c) => {
+        Value::TypeConstructor(enum_, c) => {
             return Err(EvalError::TypeError(format!(
-                "Type constructor as captured value not supported: {c}."
+                "Type constructor as captured value not supported: {}::{c}.",
+                enum_.name
             )))
         }
-        Value::Enum(variant, _items) => {
-            // The main problem is that we do not know the type of the enum.
-            return Err(EvalError::TypeError(format!(
-                "Enum as captured value not supported: {variant}."
-            )));
+        Value::Enum(enum_, variant, items) => {
+            let variant_ref = Expression::Reference(
+                SourceRef::unknown(),
+                Reference::Poly(PolynomialReference {
+                    name: format!("{}::{variant}", enum_.name),
+                    // We do not know the type args here.
+                    type_args: None,
+                }),
+            );
+            match items {
+                None => variant_ref,
+                Some(items) => FunctionCall {
+                    function: Box::new(variant_ref),
+                    arguments: items
+                        .iter()
+                        .map(|i| try_value_to_expression(i))
+                        .collect::<Result<_, _>>()?,
+                }
```

Comment:
Not directly related to this PR, but couldn't we simplify this by making the items non optional, and making `MyEnum::NoArgs` sugar for `MyEnum::NoArgs()`?

Reply:
Just checking if I understand this correctly. Let's assume we have```
enum A {
  X,
  Y(),
}
```
(setting aside the question if we want to support an enum variant of the kind `Y()` - or was that the main question?), then `A::X` is an enum value while `A::Y` as a function that returns an enum value.

What we could do is change the items in `evaluator::Value::Enum` from `Option<Vec<Arc<Self>>>` to `Vec<Arc<Self>>`, since for a fixed enum variant, we do not need to distinguish between `None` and `Some(vec![])`.

Is that what you meant?

Reply:
Getting rid of the `Option<Vec<_>>` would indeed be nice, but what I meant was:

Here if we have a variant with no arguments we return an enum value, otherwise we return a function call. I was thinking we could remove this edge case and always use `A::X()` internally to represent `A::X`. However it seems like it would be ambiguous because `A::X` would then be both a function and an enum value.

Closing